### PR TITLE
Adding Japanese game support

### LINKED ISSUE #62: Support for non-English Pokémon versions
Hi! I intend to strat streaming IronMon to my Fracophone audience, and as such I would like to be using the French version of the games. It would be even better if the tracker had translated Pokémon, abilities names…

I tried to edit the code to force the script to load with a French LeafGreen ROM, trying 1.0 and 1.1 variables, unfortunately the tracker does not work beyond detecting the Pokémon species and crashes the emulator in battle, so I assume memory addresses are different for every release.

Besides English, Pokémon FR, LG and E have been released in the following languages:
- Japanese
- French
- German
- Spanish
- Italian

I looked around at the code and I think (this maybe wrong or incomplete) implementing multi-language support would entail f all required memory addresses for all revisions of every language release (according to TCRF, the Japanese version has V1.0 and V1.1 releases, while the European languages did not receive a revision)

If we also want to go the route of adding names of Pokémon, Abilities, Moves, Items, Natures, Statuses and types in several languages (note that Spanish and Italian use English Pokémon names), they would need to be added to the data files.

I would definitely be willing to contribute to this should you decide that this is within the scope’s project. Depending on the approach for doing the second part, I could try to write automated scripts with PokéAPI to add variables with names in other languages.

If you know how to find the memory addresses, adding them to the GameSettings.lua would be trivial if not a little time-consuming, but unfortunately at the moment I wouldn’t know where to look for to find the right addresses, but this is something I can learn.

Sorry, this is very wordy! As this is something I really want to see added to the tracker, I made sure to research the feasibility, and as I said I would be willing to contribute as much as possible to realize it. I think it would help the IronMon open up to broader audiences outside of the Anglophone streamers and players. I would understand if you would consider this too heavy to support though − in that case I would certainly do my own fork that supports at least French language games.

Thanks for reading!

### PR DESCRIPTION
**Pull request by UTDZac**
This adds compatibility support to allow the Tracker to play Japanese FireRed ROMs, similar to how it can play other non-English games. This is only now possible because of the support for Japanese character rendering in Bizhawk 2.9. As for mGBA, it technically displays these text characters fine, but the formatting is still messy because there is no way to properly determine the length of a UTF16 string in this context.

The translations were imported from this [Github repo](https://github.com/roytam1/rtoss/tree/master/PokemonHackSourceCode/PokemonMemHack).

I had to do some hacking in `GameSettings` to allow for Japanese ROM addresses, mostly cause EWRAM and IWRAM were different. I want to say that 99% of the addresses are correct, but it's possible some of them may not work.

image

### CODE REVIEW

**Comment by UTDZac** [comment]
Forgot to note, the changes to `Battle.lua` were mostly to account for `nil` addresses or undefined ability addresses.

File: ironmon_tracker/GameSettings.lua [hunk 1]
```diff
@@ -401,6 +411,10 @@ function GameSettings.setGameVersion(gameversion)
 				versionName = "Pokemon Feuerrote",
 				versionIndex = 6,
 			},
+			[0x00630000] = { -- Japanese
+				versionName = "Pokemon FireRed J",
+				versionIndex = 7,
+			},
```
Comment (Fellshadow) [inline ironmon_tracker/GameSettings.lua:417]: Could be worth mentioning the `v1.0` here? (either directly in `versionName` or as a comment, just so the note is there for anyone else who looks at this)

**Review by Fellshadow — APPROVED** [review]
Seems fine enough to me, hopefully addresses work fine 👀

File: ironmon_tracker/GameSettings.lua (hunk 1 above)
Comment (UTDZac) [reply ironmon_tracker/GameSettings.lua:417]: Part of why i left this out is my lack of research to know what rom i was using and what roms are possible. I *think* the rom im using is 1.0. and I *think* its the only one that exist. but im not entirely sure. I dont have a different version that i found that i could test.

Comment (Fellshadow) [reply ironmon_tracker/GameSettings.lua:417]: hm yeah, that's fair i suppose

**Pull request merged**